Mathlib/Topology/Algebra/RestrictedProduct.lean

```diff
@@ -268,6 +282,147 @@ instance [Π i, Ring (R i)] [∀ i, SubringClass (S i) (R i)] :
+product `Πʳ i, [R i, B i]_[𝓕]` to the component `R j`."]
+def evalMonoidHom (j : ι) [Π i, Monoid (R i)] [∀ i, SubmonoidClass (S i) (R i)] :
+    (Πʳ i, [R i, B i]_[𝓕]) →* R j where
+  toFun x := x j
+  map_one' := rfl
+  map_mul' _ _ := rfl
+
+@[simp]
+lemma evalMonoidHom_apply [Π i, Monoid (R i)] [∀ i, SubmonoidClass (S i) (R i)]
+    (x : Πʳ i, [R i, B i]_[𝓕]) (j : ι) : evalMonoidHom R j x = x j :=
+  rfl
+
+/-- `RestrictedProduct.evalRingHom j` is the ring homomorphism from the restricted
+product `Πʳ i, [R i, B i]_[𝓕]` to the component `R j`.
+-/
+def evalRingHom (j : ι) [Π i, Ring (R i)] [∀ i, SubringClass (S i) (R i)] :
+    (Πʳ i, [R i, B i]_[𝓕]) →+* R j where
+  __ := evalMonoidHom R j
+  __ := evalAddMonoidHom R j
+
+@[simp]
+lemma evalRingHom_apply [Π i, Ring (R i)] [∀ i, SubringClass (S i) (R i)]
+    (x : Πʳ i, [R i, B i]_[𝓕]) (j : ι) : evalRingHom R j x = x j :=
+  rfl
+
+end eval
+
+section map
+
+variable {ι₁ ι₂ : Type*}
+variable (R₁ : ι₁ → Type*) (R₂ : ι₂ → Type*)
+variable {𝓕₁ : Filter ι₁} {𝓕₂ : Filter ι₂}
+variable {A₁ : (i : ι₁) → Set (R₁ i)} {A₂ : (i : ι₂) → Set (R₂ i)}
+variable {S₁ : ι₁ → Type*} {S₂ : ι₂ → Type*}
+variable [Π i, SetLike (S₁ i) (R₁ i)] [Π j, SetLike (S₂ j) (R₂ j)]
+variable {B₁ : Π i, S₁ i} {B₂ : Π j, S₂ j}
+variable (f : ι₂ → ι₁) (hf : Tendsto f 𝓕₂ 𝓕₁)
+
+section set
+
+variable (φ : ∀ j, R₁ (f j) → R₂ j) (hφ : ∀ᶠ j in 𝓕₂, A₁ (f j) ⊆ φ j ⁻¹' A₂ j)
+
+/--
+Given two restricted products `Πʳ (i : ι₁), [R₁ i, A₁ i]_[𝓕₁]` and `Πʳ (j : ι₂), [R₂ j, A₂ j]_[𝓕₂]`,
+`RestrictedProduct.map` gives a function between them. The data needed is a function `f : ι₂ → ι₁`
+such that `𝓕₂` tends to `𝓕₁` along `f`, and functions `φ j : R₁ (f j) → R₂ j`
+sending `A₁ (f j)` into `A₂ j` for an `𝓕₂`-large set of `j`'s.
+
+See also `mapMonoidHom`, `mapAddMonoidHom` and `mapRingHom` for variants.
+-/
+def map (x : Πʳ i, [R₁ i, A₁ i]_[𝓕₁]) : Πʳ j, [R₂ j, A₂ j]_[𝓕₂] := ⟨fun j ↦ φ j (x (f j)), by
+  filter_upwards [hf.eventually x.2, hφ] using fun _ h1 h2 ↦ h2 h1⟩
+
+lemma map_id (x : Πʳ i, [R₁ i, A₁ i]_[𝓕₁]) : map R₁ R₁ id
+  (fun ⦃_⦄ a ↦ a : Tendsto id 𝓕₁ 𝓕₁) (fun _ t ↦ t)
+  (Eventually.of_forall fun _ ↦ (fun ⦃_⦄ a ↦ a)) x = x := rfl
+
+lemma map_comp {ι₃} (R₃ : ι₃ → Type*) {𝓕₃ : Filter ι₃} {A₃ : (i : ι₃) → Set (R₃ i)} (g : ι₃ → ι₂)
+    (hg : Tendsto g 𝓕₃ 𝓕₂) (ψ : ∀ k, R₂ (g k) → R₃ k) (hψ : ∀ᶠ k in 𝓕₃, A₂ (g k) ⊆ ψ k ⁻¹' A₃ k) :
+    map R₁ R₃ (f ∘ g) (hf.comp hg) (fun k t ↦ ψ k (φ (g k) t)) (by
+      filter_upwards [hψ, hg hφ] with k h23 h12 t ht1
+      exact h23 (h12 ht1)) =
+    (map R₂ R₃ g hg ψ hψ) ∘ (map R₁ R₂ f hf φ hφ) := rfl
+
+@[simp]
+lemma map_apply (x : Πʳ i, [R₁ i, A₁ i]_[𝓕₁]) (j : ι₂) :
+    x.map R₁ R₂ f hf φ hφ j = φ j (x (f j)) :=
+  rfl
+
+end set
+
+section monoid
+
+variable [Π i, Monoid (R₁ i)] [Π i, Monoid (R₂ i)] [∀ i, SubmonoidClass (S₁ i) (R₁ i)]
+    [∀ i, SubmonoidClass (S₂ i) (R₂ i)] (φ : ∀ j, R₁ (f j) →* R₂ j)
+    (hφ : ∀ᶠ j in 𝓕₂, B₁ (f j) ≤ φ j ⁻¹' B₂ j)
```

Comment:
This `≤` should be an inclusion, right?

Reply:
I was following your convention where the A's were sets and the B's were SetLike (and in fact are submonoids here in the applications). Are you suggesting `⊆`? This doesn't compile.

Reply:
I claim that it doesn't compile for stupid reasons: in the end, this is an inequality between *sets* (so indeed an inclusion), not subobject types, precisely because of the preimage.

Reply:
I don't think there is anything specific. Note that we cannot have any analog of e.g [Subgroup.map](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Subgroup/Map.html#Subgroup.map) for any `SubgroupClass` type, because group homomorphisms do not map vector subspaces to vector subspaces.

Reply:
I'll just use Set.MapsTo again